### PR TITLE
ipq40xx: Add support for Linksys MR6350

### PR DESCRIPTION
**Pull request by RolandoMagico**
This pull request is based on
- the discussions in https://forum.openwrt.org/t/adding-openwrt-support-for-linksys-mr6350
- https://github.com/openwrt/openwrt/pull/11405 which added support for similar devices.

Device Specs:
- IPQ4019
- Quad Core CPU
- 256 MB RAM
- 256 MB FLASH
- 4 LAN ports, 1 WAN port
- 2.4GHz (802.11n) and 5GHz (802.11c) wifi
- 3 LEDs (Red, blue, green) which are routed to one indicator at the top of the case
- 2 buttons (Reset, WPS)

Disassembling the device:
- There are 4 screws at the bottom of the device which must be removed
- Two are under the front rubber feets
- Two are under the labels in the back (corner next to the rear rubber feets)

Serial interface:
- The serial interface is already populated on the device with a 6-pin header
- Pin 1 is next to the heatsink
- Pinout: 1: 3.3V, 2: TX, 3: RX, 4: unknown, 5: GND, 6: GND
- Settings: 115200, 8N1

Migrating to OpenWrt requires multiple steps:
- Load and boot the initramfs image
- Adapt U-Boot settings to support bigger kernels
- Flash the sysupgrade image

Load and boot initramfs:
- Connect serial interface
- Set up a TFTP server on IP 192.168.1.254
- Copy openwrt-ipq40xx-generic-linksys_mr6350-initramfs-zImage.itb to TFTP server
- Rename file to C0A80101.img
- Boot up the device and stop in U-Boot
- Run the following U-Boot commands after a link has been established: 
```tftp```
```bootm```
- Initramfs image is started now.

Adapt U-Boot settings to support bigger kernels:
- Run "fw_printenv" in the initramfs image  after booting
- There should be an entry kernsize=300000 which indicates the maximum size for the kernel is 3MB
- Execute "fw_setenv kernsize 500000" to increase the max kernel size to 5MB
- Check that the change are applied with "fw_printenv"

Flash the sysupgrade image:
- Default sysupgrade routine either with a initramfs image containing LuCI or via command line.

Revert back to OEM firmware:
- Only tested with FW_MR6350_1.1.3.210129_prod.img
- Flash the OEM firmware via sysupgrade
- Forced update is required

### CODE REVIEW

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		led_green: green {
+			label = "green:panel";
+			gpios = <&tlmm 49 GPIO_ACTIVE_HIGH>;
+		};
+
+
+		// On back panel, above USB socket
+
+		led_usb: usb {
+			label = "green:usb";
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:58]:
```suggestion
		led-usb {
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_USB;
```
(label is not used anywhere)

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			trigger-sources = <&usb3_port1>, <&usb3_port2>,
+					  <&usb2_port1>;
+			linux,default-trigger = "usbport";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:69]:
```suggestion
		button-reset {

```

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&tlmm 50 GPIO_ACTIVE_LOW>;
+		};
+
+		wps {
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:75]:
```suggestion
		button-wps {

```

File: target/linux/ipq40xx/image/generic.mk [hunk 4]
```diff
@@ -705,6 +705,23 @@ define Device/linksys_ea8300
...
+	DEVICE_VENDOR := Linksys
+	DEVICE_MODEL := MR6350
+	SOC := qcom-ipq4019
+	KERNEL_SIZE := 5120k
+	IMAGE_SIZE := 84992k
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	UBINIZE_OPTS := -E 5    # EOD marks to "hide" factory sig at EOF
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=MR6350
```
Comment (raenye) [inline target/linux/ipq40xx/image/generic.mk:720]:
```suggestion
	IMAGE/factory.bin  := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | linksys-image type=MR6350
```
Why double-escape?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		led-running = &led_green;
+		led-upgrade = &led_blue;
+		serial0 = &blsp1_uart1;
+	};
+
+	// Top panel LEDs, above Linksys logo
+	leds {
+		compatible = "gpio-leds";
+		
+		led_blue: blue {
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:38]:
```suggestion
		led_blue: led-blue {

```

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			label = "blue:panel";
+			gpios = <&tlmm 46 GPIO_ACTIVE_LOW>;
+		};
+
+		led_red: red {
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:43]:
```suggestion
		led_red: led-red {
```

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+			label = "red:panel";
+			gpios = <&tlmm 47 GPIO_ACTIVE_HIGH>;
+			panic-indicator;
+		};
+
+		led_green: green {
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:49]:
```suggestion
		led_green: led-green {
```

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		led-upgrade = &led_blue;
+		serial0 = &blsp1_uart1;
+	};
+
+	// Top panel LEDs, above Linksys logo
+	leds {
+		compatible = "gpio-leds";
+		
+		led_blue: blue {
+			label = "blue:panel";
```
Comment (raenye) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts:39]: `panel` is not a known function. I wonder if it makes sense to use one (`status`?)

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 1 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: @raenye: Thanks for the review. I fixed the findings

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 2 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: Fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 3 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: fixed

File: target/linux/ipq40xx/image/generic.mk (hunk 4 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/image/generic.mk]: fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 5 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 6 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 7 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts (hunk 8 above)
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+		
+		led_blue: led-blue {
+			label = "blue:status";             // Can be removed as soon as https://github.com/openwrt/openwrt/pull/13837 is merged
+			color = <LED_COLOR_ID_BLUE>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&tlmm 46 GPIO_ACTIVE_LOW>;
+		};
+
+		led_red: led-red {
+			label = "red:status";              // Can be removed as soon as https://github.com/openwrt/openwrt/pull/13837 is merged
```
[the remark this answers is not in the document]
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: Thanks, I removed the line from the DTS and LEDs are still working as expected

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+			color = <LED_COLOR_ID_RED>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&tlmm 47 GPIO_ACTIVE_HIGH>;
+			panic-indicator;
+		};
+
+		led_green: led-green {
+			label = "green:status";            // Can be removed as soon as https://github.com/openwrt/openwrt/pull/13837 is merged
```
[the remark this answers is not in the document]
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: Updated

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		led-upgrade = &led_blue;
+		serial0 = &blsp1_uart1;
+	};
+
+	// Top panel LEDs, above Linksys logo
+	leds {
+		compatible = "gpio-leds";
+		
+		led_blue: led-blue {
+			label = "blue:status";             // Can be removed as soon as https://github.com/openwrt/openwrt/pull/13837 is merged
```
[the remark this answers is not in the document]
Comment (RolandoMagico) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mr6350.dts]: Done

**Comment by RolandoMagico** [comment]
I removed the WIP flag because wireless is also working quite well with the "default" calibration data. Someone has to decide if this PR can be merged without https://github.com/openwrt/firmware_qca-wireless/pull/24 or if we have to wait for the firmware_qca-wireless PR.

**Comment by Djfe** [comment]
why did you close this PR?

**Comment by RolandoMagico** [comment]
@Djfe: I don’t use the hardware anymore, so I cannot test any further changes.
See also https://forum.openwrt.org/t/adding-openwrt-support-for-linksys-mr6350/82812/177

**Comment by RolandoMagico** [comment]
@raenye, @Djfe:I got the device again, tested based on latest main but cannot reopen the PR. Anyone has rights to do it?

**Comment by hauke** [comment]
I can not reopen it. github complains that the branch was force pushed or recreated.
Just open a new one and link this PR.

**Pull request closed without merging**